File: Linked List Cycle - Leetcode 141/Linked List Cycle - Leetcode 141.java
class ListNode {
    int val;
    ListNode next;

    ListNode(){}    
    ListNode(int val){
        this.val = val;
    }
    ListNode(int val, ListNode next)
    {
        this.val = val;
        this.next = next;
    }
}

class Solution {
    public ListNode hasCycle(ListNode head)
    {
        ListNode c = head;
        Set<ListNode> s = new HashSet<>();

        while (c !== null)
        {
            if (s.contains(c))
            {
                return true
            } else {
                s.add(c):
                c = c.next;
            }
        }
    return false;
    }

    public ListNode hasCycleOptimize(ListNode head)
    {
        ListNode d = new ListNode();
        d.next = head;
        ListNode s = d, f = d;

        while( f && f.next)
        {
            if (s === f)
            {
                return true;
            }
            s = s.next;
            f = f.next.next;
        }

        return false;

    }
}

File: Linked List Cycle - Leetcode 141/Linked List Cycle - Leetcode 141.ts
class ListNode {
  public val: number;
  public next: ListNode | null;

  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

// Solution 1: O(n), T(n)
const hasCycle = (head: ListNode | null) => {
  const s = new Set();
  let c = head;

  while (c !== null) {
    if (s.has(c)) {
      return true;
    } else {
      s.add(c);
      c = c.next;
    }
  }
  return false;
};

// Solution 2: T(n), S(1)
const hasCycleOptimize = (head: ListNode | null) => {
  const d = new ListNode();
  d.next = head;
  let s = d,
    f = d;
  while (f && f.next) {
    if (s === f) {
      return true;
    }
    s = s.next!;
    f = f.next.next!;
  }
  return false;
};
